improv_ble_client/improv_ble_client.py

```diff
@@ -0,0 +1,469 @@
+        self.loop = asyncio.get_running_loop()
+
+    def set_ble_device_and_advertisement_data(
+        self, ble_device: BLEDevice, advertisement_data: AdvertisementData
+    ) -> None:
+        """Set the ble device."""
+        self._ble_device = ble_device
+        self._advertisement_data = advertisement_data
+
+    @property
+    def address(self) -> str:
+        """Get the address of the device."""
+        return str(self._ble_device.address)
+
+    @property
+    def name(self) -> str:
+        """Get the name of the device."""
+        return str(self._ble_device.name or self._ble_device.address)
+
+    @property
+    def rssi(self) -> int | None:
+        """Get the rssi of the device."""
+        if self._advertisement_data:
+            return self._advertisement_data.rssi
+        return None
+
+    async def can_identify(self) -> bool:
+        """Return if the device supports identify."""
+        _LOGGER.debug("%s: can_identify", self.name)
+
+        async def _can_identify() -> bool:
+            capabilities = await self.read_characteristic(
+                CHARACTERISTIC_UUID_CAPABILITIES
+            )
+            return bool(capabilities & prot.Capabilities.IDENTIFY)
+
+        return await self._execute(_can_identify)
+
+    async def identify(self) -> None:
+        """Identify the device."""
+        _LOGGER.debug("%s: identify", self.name)
+
+        async def _identify() -> None:
+            await self.send_cmd(prot.IdentifyCmd())
+
+        await self._execute(_identify)
+
+    async def need_authorization(self) -> bool:
+        """Return if the device needs authorization."""
+        _LOGGER.debug("%s: need_authorization", self.name)
+
+        async def _need_authorization() -> bool:
+            state = await self.read_characteristic(CHARACTERISTIC_UUID_STATE)
+            return state == prot.State.AUTHORIZATION_REQUIRED
+
+        return await self._execute(_need_authorization)
+
+    async def provision(
+        self,
+        ssid: str,
+        password: str,
+        state_callback: Callable[[prot.State], None] | None,
+    ) -> str | None:
+        """Provision the device.
+
+        Returns the redirect url or None.
+        """
+        _LOGGER.debug("%s: provision ssid: %s, pw: %s", self.name, ssid, password)
+
+        async def _provision() -> str | None:
+            """Execute the procedure"""
+
+            def handle_error(value: prot.Error) -> None:
+                error_fut.set_result(value)
+
+            def handle_state(state: prot.State) -> None:
+                if state_callback:
+                    state_callback(state)
+
+            subscriptions = [
+                self._notification_handlers.subscribe_error(handle_error),
+                self._notification_handlers.subscribe_state(handle_state),
+            ]
+            error_fut: asyncio.Future[prot.Error] = asyncio.Future()
+            provisioned_fut = self.receive_response(prot.WiFiSettingsRes)
+
+            try:
```

Comment:
Does this need a timeout?  I may be easier to handle cancelation inside the inner functions on timeout if you limit the scope of the timeout to this function

Reply:
To not simplify things I chose to only have a single global timeout, handled by the `ImprovBLEClient._disconnect_timer`.
Could you give an example of how cancellation could be done if the timeout scope is narrowed?

Reply:
You have to make sure all the code running under the timeout is cancellation safe (ie cleans up and disconnects ok if it gets canceled)

By limiting the scope of where we expect cancellation to happen to just the area where its going to timeout thats a bit easier, but its also better to have all the code handle cancellation

Reply:
Right, I prefer to handle the the cancellation at one place higher up instead of in several places further down.

Reply:
Its probably safe since there is a disconnect timer that won't be canceled when the task is cancelled.

Maybe it would be good to add a test to make sure if any operation is cancelled by the task its running in being canceled, it still disconnects.